interp/notation.ml

```diff
@@ -1255,10 +1256,51 @@ type prim_token_infos = {
 let prim_token_interp_infos =
   ref (String.Map.empty : (required_module * prim_token_interp_info) String.Map.t)
 
-(* Table from global_reference to backtrack-able informations about
+(* Backtrackable state for uninterpretation information.
+   This is the state that enables disabling notations.
+ *)
+type prim_token_uninterp_info =
+  { pti_scope : scope_name ;
+    pti_info : prim_token_interp_info ;
+    pti_in_match : bool ;
+    pti_enabled : bool }
+
+(* Table from global_reference to backtrack-able information about
    prim_token uninterpretation (in particular uninterpreter unique id). *)
 let prim_token_uninterp_infos =
-  ref (GlobRef.Map.empty : ((scope_name * (prim_token_interp_info * bool)) list) GlobRef.Map.t)
+  ref (GlobRef.Map.empty : prim_token_uninterp_info list GlobRef.Map.t)
+
+let set_prim_notation_status : env:Environ.env -> gref:GlobRef.t -> ?scopes:scope_name list -> is_string:bool -> bool -> int =
+  fun ~env ~gref ?scopes ~is_string enable ->
+    let matches ty (info : prim_token_uninterp_info) : bool =
+      let check_scope =
+        match scopes with
+        | None -> true
+        | Some scp -> List.mem info.pti_scope scp
+      in
+      let check_ref () =
+        let open Environ in
+        QGlobRef.equal env gref ty ||
+        match info.pti_info with
+        | NumberNotation info when not is_string ->
+          (QGlobRef.equal env gref info.to_ty ||
+           QGlobRef.equal env gref info.of_ty)
+        | StringNotation info when is_string ->
+          (QGlobRef.equal env gref info.to_ty ||
+           QGlobRef.equal env gref info.of_ty)
+        | _ -> false
+      in check_scope && check_ref ()
+    in
+    let cnt = ref 0 in
+    prim_token_uninterp_infos :=
+      GlobRef.Map.mapi (fun ty ->
+        (List.map (fun info ->
+             if matches ty info
+             then
+               let _ = cnt := 1 + !cnt in
+               { info with pti_enabled = enable }
```

Comment:
By comparison, `Disable Notation` does not disable (resp. enable) notations which are already disabled (resp. enabled). Instead, it fails. I don't have strong opinion on what policy to adopt, but a consistent policy would be nice.